Feature/error handling

dscottboggs opened the pull request:
This will resolve #15

dscottboggs committed:
add fixtures file in place of permissions.json

dscottboggs committed:
initial progress on error handlers

dscottboggs committed:
Made fixtures less magic

dscottboggs committed:
replace macro with constant

dscottboggs committed:
polish and revise

the CLI spec no longer needs the with_state_restore;
moved Fixtures to the top-level;
removed reopening of Config in Fixtures

dscottboggs committed:
Only handle error statuses, as opposed to all statuses

dscottboggs committed:
Update specs to reflect new error handlers

dscottboggs committed:
only run the server once

j8r commented:
`Exception#cause` can be used, like in https://github.com/DFabric/dppm/blob/master/src/logger.cr#L48, to implement https://jsonapi.org/examples/#error-objects-multiple-errors

dscottboggs commented:
Interesting... I'll get started on this feature.

Shouldn't [this](https://github.com/DFabric/dppm/blob/4e0eac5b81274fbccd9ed3f94afc67f34fa13a4a/src/logger.cr#L49) be `print_error ex.message`?

j8r commented:
In fact, [it's the same](https://github.com/crystal-lang/crystal/blob/master/src/exception.cr#L42).
Either way there would be an Union, message being nilable.

j8r commented on spec/spec_helper.cr:
```diff
@@ -22,7 +34,7 @@ PERMISSION_FILE = Path[__DIR__, "permissions.json"]
   DppmRestApi.permissions_config = Fixtures.permissions_config
   DppmRestApi.permissions_config.write_to PERMISSION_FILE
-  DppmRestApi.run Socket::IPAddress::LOOPBACK, DPPM::Prefix.default_dppm_config.port, __DIR__
+  run_server unless already_running
```
Why? Also, a method isn't needed.

dscottboggs replied:
> Why?

Because you only need to start the server once. Starting it multiple times may have unintended side effects. I thought it best to only start it once.

> a method isn't needed.

No, but it makes it neater and more readable, hence more maintainable in the future.

j8r replied:
Ok. For the method, I don't think so, because we have to look back at another abstraction to see what is the actual code ran.
A method is useful if it's used multiple times.

But, why all of this instead of just having this call before `Spec.before_each`, as done first by me?

dscottboggs replied:
so option A we can do

```crystal
already_running = false
Spec.before_each do
  DppmRestApi.permissions_config = Fixtures.permissions_config
  DppmRestApi.permissions_config.write_to PERMISSION_FILE
  DppmRestApi.run Socket::IPAddress::LOOPBACK, DPPM::Prefix.default_dppm_config.port, __DIR__ unless already_running
  already_running = true
end
```
or, alternatively --

```crystal
DppmRestApi.permissions_config = Fixtures.permissions_config
DppmRestApi.permissions_config.write_to PERMISSION_FILE
DppmRestApi.run Socket::IPAddress::LOOPBACK, DPPM::Prefix.default_dppm_config.port, __DIR__
Spec.before_each do
  DppmRestApi.permissions_config = Fixtures.permissions_config
  DppmRestApi.permissions_config.write_to PERMISSION_FILE
end
```
I didn't like the repetitiveness of the latter solution, that was my rationale for the first. The method was added because the last line of the before_each block is very long

dscottboggs replied:
We could use the latter option, but with less repetition like

```crystal
def reset_config
  DppmRestApi.permissions_config = Fixtures.permissions_config
  DppmRestApi.permissions_config.write_to PERMISSION_FILE
end

reset_config
DppmRestApi.run Socket::IPAddress::LOOPBACK, DPPM::Prefix.default_dppm_config.port, __DIR__

Spec.before_each { reset_config }

j8r replied:
The latter option sounds good, one less conditional and boolean 👍

j8r commented on src/actions/app.cr:
```diff
@@ -48,28 +44,27 @@ module DppmRestApi::Actions::App
       elsif config = app.get_config(key)
-        context.response.content_type = "application/json"
-        {"data" => config, "errors" => [] of Nil}.to_json context.response
+        {"data" => config}.to_json context.response
```
Can be a `NamedTuple` (more efficient than `Hash`)

j8r commented on src/config/jwt_compatible_hash.cr:
```diff
@@ -0,0 +1 @@
+alias JWTCompatibleHash = Hash(String, String | Int32 | Bool?)
```
A file with a single alias? Why it isn't put in an already existing module, like `DppmRestApi`?

dscottboggs replied:
I was trying to think of a place to put it where it fit with other stuff, and there just didn't seem to be any. It doesn't seem like a good idea to clutter the main file with stuff like this, and putting it anywhere else would make it hard to find. It doesn't do any harm being there. We could put it back in `src/utils.cr` but you seemed very much against the idea of having that at all.

j8r replied:
You need it at `src/config/user.cr`, so it can be defined there.

dscottboggs replied:
Ok, but if we find a reason to use that type outside of User then we should move it from there, but for now that looks fine.

dscottboggs committed:
add ability to handle multiple errors through Exception#cause
make errors list a list of objects to allow passing more information through later on

j8r commented on spec/actions/app_spec.cr:
```diff
@@ -7,84 +7,140 @@ module DppmRestApi::Actions::App
       it "responds with 401 Unauthorized" do
         get fmt_route "/some-app/config/key"
         response.status_code.should eq 401
+        ErrorResponse.from_json(response.body)
```
All this lines can be a a helper method, like `assert_error_response(response)`

j8r replied:
(inspired by https://github.com/crystal-lang/crystal/blob/master/spec/std/json/pull_parser_spec.cr#L88)

dscottboggs replied:
Implemented.

dscottboggs commented:
That is a good idea

dscottboggs committed:
add helper method to clean up specs

j8r commented on spec/spec_helper.cr:
```diff
@@ -24,6 +24,14 @@ struct ErrorResponse
   property errors : Array(ErrorData)
 end
 
+def assert_unauthorized!(response : HTTP::Client::Response)
```
I don't get why there is an exclamation mark. There are usually used to mark a method as "dangerous", like `not_nil!`.

dscottboggs replied:
This is dangerous method, it raises an exception to fail a test. Regardless, changed.

j8r commented on src/middlewares/error_handlers.cr:
```diff
@@ -1,16 +1,44 @@
+      {
+        type:    error.class.to_s,
+        message: error.message,
+      }
```
Not a fan of returning a `NamedTuple`. I suggest a record, with `status_code` being nillable.
You can use the `ErrorData` from the `spec_helper` and put it there.

dscottboggs replied:
Actually, I just moved the structs from spec_helper into that file, added a couple methods on them, and everything is neater. A nicer logging solution exposed itself in the process too :)

I don't like the `record` macro though, it seems silly, as it takes very little more to type out a `struct` with `property`s on it. It just seems unnecessary.

dscottboggs committed:
use structs instead of NamedTuple for error responses

j8r commented on spec/spec_helper.cr:
```diff
@@ -39,4 +39,4 @@ DppmRestApi.run Socket::IPAddress::LOOPBACK, DPPM::Prefix.default_dppm_config.po
 Spec.before_each { reset_config }
 # Clean up after ourselves
-Spec.after_each { File.delete PERMISSION_FILE if File.exists? PERMISSION_FILE }
+Spec.after_each { File.delete PERMISSION_FILE if File.exists? PERMISSION_FILE }
```
Please remove this commit

j8r commented on src/middlewares/error_handlers.cr:
```diff
@@ -1,29 +1,59 @@
   end
 
-  def response_data_for(error : Exception)
+  def self.from(error : Exception)
```
`from_exception` would be a better name, to differentiate with `from_json`

dscottboggs replied:
With JSON, one usually has an IO or file or something, so one calls `ErrorResponse.from_json(context.response)` or `ErrorResponse.from_json file`. `#from_exception` would stutter with a variable named `exception`, like in the block where it's called from:

```crystal
response_data = ErrorResponse.from_exception exception
```

I *really* don't like that. Can we compromise on a `self.new` overload?

j8r replied:
Good idea yes

j8r commented on src/middlewares/error_handlers.cr:
```diff
@@ -1,16 +1,74 @@
+  property message : String?
+  property status_code : HTTP::Status?
+
+  def initialize(@type, @message, @status_code = nil)
```
```suggestion
  def initialize(@type : String, @message : String?, @status_code : HTTP::Status? = nil)
```

dscottboggs replied:
What is the point of this? A compiler error will still result if one calls `ErrorMessage.new` with a `Int32` as the first value. In fact, I had to make some changes to the types of this as I was developing it -- I originally had it like

```crystal
  property type : String
  property message : String
  property status_code : Int32?
```
and the compiler told me I couldn't call `ErrorMessage.new(String, String?, HTTP::Status?)`, as one would expect.

j8r replied:
They will appear in the API docs. Maybe the compiler error is better, too.

j8r commented on src/middlewares/error_handlers.cr:
```diff
@@ -1,16 +1,74 @@
+  include JSON::Serializable
+  property errors : Array(ErrorData)
+
+  def initialize(@errors)
```
```suggestion
  def initialize(@errors : Array(ErrorData))
```

j8r commented on src/middlewares/error_handlers.cr:
```diff
@@ -1,16 +1,74 @@
+  end
+
+  def self.from(error : Exception)
+    if error.responds_to? :status_code
```
I think it's possible to avoid this, by adding a method overload, more restrictive than `Exception`

dscottboggs replied:
updated

dscottboggs replied:
I wish that worked... :slightly_frowning_face:

dscottboggs committed:
use type restrictions instead of responds_to?

rename ErrorResponse.from and ErrorData.from to .new

dscottboggs committed:
added more type restrictions

j8r commented on src/errors/http_status_errors.cr:
```diff
@@ -0,0 +1,30 @@
+
+    def self.status_code
+      STATUS_CODE
+    end
```
Why?

dscottboggs replied:
Why not?

j8r replied:
Because one can already use `STATUS_CODE`. Unless you have something against uppercase letters?

j8r commented on src/actions/pkg.cr:
```diff
@@ -1,3 +1,5 @@
+require "../actions"
```
Not needed

dscottboggs replied:
My editor hooks in with the compiler to alert me to issues with types or syntax while I'm working, and that feature is only effective if each file has the appropriate `require` statements. It doesn't hurt anything to be there.

j8r replied:
This is a non sense, this file is required by a file (`actions.cr`), that is also required back?!
I don't know about your editor, but this isn't a strong enough argument against this anti-patterns.

Nobody wants to see this unneeded backward parent requirements everywhere.

dscottboggs replied:
:open_mouth: wow, if it's that important to you...

j8r replied:
Yes it is, I don't invent anything there, I'm looking at how other big projects like crystal, shards, lucky and other non-crystal projects.

That's a modularization principle: why here `pkg` will require `actions`, that will bring `src`, `app` etc?

Only import/require only what is needed – simpler to understand, avoid circle dependencies.

j8r commented on src/actions/app.cr:
```diff
@@ -1,5 +1,5 @@
 require "dppm"
-# require "../config/route"
+require "../actions"
```
Not needed

j8r commented on src/actions/src.cr:
```diff
@@ -1,18 +1,20 @@
+require "../actions"
```
Not needed

j8r commented on src/actions/service.cr:
```diff
@@ -1,3 +1,5 @@
+require "../actions"
```
Not needed

j8r commented on src/middlewares/error_handlers.cr:
```diff
@@ -1,16 +1,76 @@
+  end
+
+  def self.new(error : Exception)
+    if error.is_a? DppmRestApi::HTTPStatusError
```
This doesn't work as a method overload (`def self.new(error : DppmRestApi::HTTPStatusError)`)?

dscottboggs replied:
Yeah, I just double-checked and indeed that does not work because `DppmRestApi::HTTPStatusError` inherits from `Exception`

dscottboggs commented:
Crap, I thought it didn't but it might have been due to a mistake elsewhere let me double check.
